File: crop-me/src/main/java/com/mikelau/croperino/Croperino.java
package com.mikelau.croperino;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.DialogInterface;
import android.content.Intent;
import android.hardware.camera2.CameraAccessException;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.v4.content.FileProvider;
import android.util.Log;

import com.mikelau.magictoast.MagicToast;

import java.io.File;
import java.io.IOException;

/**
 * Created by Mike on 9/15/2016.
 */
public class Croperino {
    private static String TAG = Croperino.class.getSimpleName();

    public static void runCropImage(File file, Activity ctx, boolean isScalable, int aspectX, int aspectY, int color, int bgColor) {
        Intent intent = new Intent(ctx, CropImage.class);
        intent.putExtra(CropImage.IMAGE_PATH, file.getPath());
        intent.putExtra(CropImage.SCALE, isScalable);
        intent.putExtra(CropImage.ASPECT_X, aspectX);
        intent.putExtra(CropImage.ASPECT_Y, aspectY);
        intent.putExtra("color", color);
        intent.putExtra("bgColor", bgColor);
        ctx.startActivityForResult(intent, CroperinoConfig.REQUEST_CROP_PHOTO);
    }

    public static void prepareChooser(final Activity ctx, String message, int color) {
        CameraDialog.getConfirmDialog(ctx, ctx.getResources().getString(R.string.app_name),
                message,
                "CAMERA",
                "GALLERY",
                "CLOSE",
                color,
                true,
                new AlertInterface.WithNeutral() {
                    @Override
                    public void PositiveMethod(final DialogInterface dialog, final int id) {
                        if (CroperinoFileUtil.verifyCameraPermissions(ctx)) {
                            prepareCamera(ctx);
                        }
                    }

                    @Override
                    public void NeutralMethod(final DialogInterface dialog, final int id) {
                        if (CroperinoFileUtil.verifyStoragePermissions(ctx)) {
                            prepareGallery(ctx);
                        }
                    }

                    @Override
                    public void NegativeMethod(final DialogInterface dialog, final int id) {

                    }
                });
    }

    public static void prepareCamera(Activity ctx) {
        try {
            Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
            Uri mImageCaptureUri;
            String state = Environment.getExternalStorageState();
            if (Environment.MEDIA_MOUNTED.equals(state)) {
                if (Uri.fromFile(CroperinoFileUtil.newCameraFile()) != null && !Uri.EMPTY.equals(CroperinoFileUtil.newCameraFile())) {
                    mImageCaptureUri = Uri.fromFile(CroperinoFileUtil.newCameraFile());
                } else {
                    mImageCaptureUri = FileProvider.getUriForFile(ctx,
                            ctx.getApplicationContext().getPackageName() + ".provider",
                            CroperinoFileUtil.newCameraFile());
                }
            } else {
                mImageCaptureUri = InternalStorageContentProvider.CONTENT_URI;
            }
            intent.putExtra(android.provider.MediaStore.EXTRA_OUTPUT, mImageCaptureUri);
            intent.putExtra("return-data", true);
            ctx.startActivityForResult(intent, CroperinoConfig.REQUEST_TAKE_PHOTO);
        } catch (Exception e) {
            if (e instanceof ActivityNotFoundException) {
                MagicToast.showError(ctx, "Activity not found.");
            } else if (e instanceof IOException) {
                MagicToast.showError(ctx, "Image file captured not found.");
            } else if (e instanceof CameraAccessException) {
                MagicToast.showError(ctx, "Camera access was denied.");
            } else {
                MagicToast.showError(ctx, "Camera access failed.");
            }
            Log.e(TAG, "Failed to prepare camera: ", e);
        }
    }

    public static void prepareGallery(Activity ctx) {
        Intent i = new Intent(Intent.ACTION_PICK, android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        ctx.startActivityForResult(i, CroperinoConfig.REQUEST_PICK_FILE);
    }
}

File: croperino/src/main/java/com/mikelau/croperino/Croperino.kt
package com.mikelau.croperino

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.provider.MediaStore
import android.widget.Toast
import androidx.core.content.FileProvider
import com.mikelau.croperino.CameraDialog.getConfirmDialog
import java.io.File
import java.io.IOException

object Croperino {

    fun runCropImage(
        file: File,
        ctx: Activity,
        isScalable: Boolean,
        aspectX: Int,
        aspectY: Int,
        color: Int,
        bgColor: Int
    ) {
        val intent = Intent(ctx, CropImage::class.java)
        intent.putExtra(CropImage.IMAGE_PATH, file.path)
        intent.putExtra(CropImage.SCALE, isScalable)
        intent.putExtra(CropImage.ASPECT_X, aspectX)
        intent.putExtra(CropImage.ASPECT_Y, aspectY)
        intent.putExtra("color", color)
        intent.putExtra("bgColor", bgColor)
        ctx.startActivityForResult(intent, CroperinoConfig.REQUEST_CROP_PHOTO)
    }

    fun prepareChooser(ctx: Activity, message: String = "", color: Int) {
        getConfirmDialog(ctx,
            ctx.resources.getString(R.string.app_name),
            message,
            "CAMERA",
            "GALLERY",
            "CLOSE",
            color,
            true,
            object : WithNeutral {
                override fun PositiveMethod(dialog: DialogInterface?, id: Int) {
                    if (CroperinoFileUtil.verifyCameraPermissions(ctx)) {
                        prepareCamera(ctx)
                    }
                }

                override fun NeutralMethod(dialog: DialogInterface?, id: Int) {
                    prepareGallery(ctx)
                }

                override fun NegativeMethod(dialog: DialogInterface?, id: Int) {}
            })
    }

    fun prepareCamera(ctx: Activity) {
        try {
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            val mImageCaptureUri: Uri
            val state = Environment.getExternalStorageState()
            mImageCaptureUri = if (Environment.MEDIA_MOUNTED == state) {
                if (Uri.fromFile(CroperinoFileUtil.newCameraFile()) != null) {
                    FileProvider.getUriForFile(
                        ctx,
                        ctx.applicationContext.packageName + ".provider",
                        CroperinoFileUtil.newCameraFile()!!
                    )
                } else {
                    FileProvider.getUriForFile(
                        ctx,
                        ctx.applicationContext.packageName + ".provider",
                        CroperinoFileUtil.newCameraFile()!!
                    )
                }
            } else {
                InternalStorageContentProvider.CONTENT_URI
            }
            intent.putExtra(MediaStore.EXTRA_OUTPUT, mImageCaptureUri)
            intent.putExtra("return-data", true)
            ctx.startActivityForResult(intent, CroperinoConfig.REQUEST_TAKE_PHOTO)
        } catch (e: Exception) {
            when (e) {
                is ActivityNotFoundException -> Toast.makeText(ctx, "Activity not found", Toast.LENGTH_SHORT).show()
                is IOException -> Toast.makeText(ctx, "Image file captured not found", Toast.LENGTH_SHORT).show()
                else -> Toast.makeText(ctx, "Camera access failed", Toast.LENGTH_SHORT).show()
            }
        }
    }

    fun prepareGallery(ctx: Activity) {
        val i = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        ctx.startActivityForResult(i, CroperinoConfig.REQUEST_PICK_FILE)
    }
}
